### ISSUE TITLE
New example: Subclassing (Advanced)

### ISSUE THREAD

**Issue by jamespec** [Team4567/Processing#6]

**Comment by jamespec** [comment]
I just pushed a new example that illustrated Java "subclassing".  This is a very advanced topic so don't worry if you don't get this.  For those familiar with writing classes, this will be interesting.  Review and let me know if you have any questions.

**Comment by jamespec** [comment]
Ugh, would be nice to mention the name!
It is called "extends", the keyword Java uses to extend (or subclass) a class.

**Comment by jabbate19** [comment]
Questions:

extends- lines 54-65
So is c a private? I tested what you said about not being usable outside of the brackets and I understand that part. 
Private means that it is only available to where it is needed in its specific class?
and Public means it is open to everyone, public I understand

ColoredBall- lines 10-21
I see where this can be helpful, such as if we are dealing with pi or irrational or complicated numbers. But where is this comment section referring to in the code?

**Comment by jamespec** [comment]
extends-lines 54-65
No, 'c' is referred to as a "local" variable.  It is declared within a block to be used only within that block.
Public and private are different.  These are added to the declarations of variable (and methods) in a class definition. 
private says that the variable can only be accessed by code that appears in the methods of this class, or, for a procedure, it can only be called by code in this class.
See lines 27,28 in Ball and 23 in ColoredBall.
These are declarations of Class Variables (sometimes called member variables).
This is where you use the public, private or protected key words.

ColoredBall lines 10-21 - Good point, I didn't actually take my own advice :)
See line 48, this could have been:
ellipse( getX(), getY(), 30, 30 );

I know that this code: ellipse( x, y, 30, 30 );
is okay because if x and y weren't usable in a subclass I would have made then private and forced the use of getX() and getY().

**Comment by jamespec** [comment]
BTW, for the lazy evaluation, it really is not an issue of how complicated a number is, PI is a constant so you can just type it in.  The issue is how much work (time) it takes to compute a number.
If you have to go off and run an algorithm for 10 mins you will want to try to avoid this.  So you don't do that calculation until somebody asks for it.

the getX() code might look something like this:

class Ball {
  float x = Float.NaN;  // Special value that means Not A Number

  float getX() {
    if(  Float.isNaN(x) ) {
        x = <go off and calculate x for a while>
    }
    return x;
  }
}

**Comment by jabbate19** [comment]
Ah ok. I know how this would help robotics. 
(All referring to Power Up 2018 and the auto from last year)
There could be a sideStart class which has void sameSide and void oppSide, called in the main 
 robot file depending if our side of the switch is on the same side we started on. 

Classes leftStart and rightStart can extend sideStart, since both would share the oppSide (move forward past the auto line and wait) and have different sameSide (turn towards the switch to try scoring)
Hopefully this year we can get the robot doing more, especially when we get encoders and gyros, all this advanced stuff will really help.

(Wrote this as you posted about lazy evaluation)
Ok, I don't think that will be an issue for the robot (except for maybe vision), but will be helpful in real-world programming I'm sure.

**Comment by jamespec** [comment]
In the sample above, if you just accessed x as b.x you would have gotten the Float.NaN not the real x.  Accessing x as getX() would calculate and return the correct x.  If this is being done than x should be private to make sure that no other code tries to access it directly.

**Comment by jamespec** [comment]
Yeah, I just added it because it is something you think about when you subclass.  Especially if you are subclassing a Class that you don't own, as we do with the WPILIB classes.

**Comment by jabbate19** [comment]
I saw in other classes there was extends and implements, whats implements for?

**Comment by jamespec** [comment]
implements is very similar but it isn't a Class you are extending but an "interface" that you are implementing.  You can say "The interface Drawable must declare a method called draw".  You don't give code you are declare that the routine must exist.  So, when you write a class that has the expected method draw you can declare the Class as "implementing Drawable".

I'll write up an example.  We will most certainly see this in the WPILIB code.
It is more common than extending a Class.

BTW, you have probably noticed that I switch back and forth on "subclassing" and "extending".
In the old days of C++ we used the words inheritence, parent class and subclass.  A subclass inherits from its parent class.  This was very confusing because you'd think the a subclass was only a portion of the parent class (sub).  So, Java tried to fix this by using the word "extends".  Everybody agrees this is much better but calling the extended class a subclass is a bad habit to break.  In fact, I'm not sure what to call it other than a subclass.
Quick summary:  
ColoredBall extends Ball
ColoredBall is a subclass of Ball
Ball is the parent or super class of ColoredBall

You can say:
ColoredBall cb = new ColoredBall( 1, 2, 3);
Ball b = cb;

but you can not say:
Ball b = new Ball( 1, 2 );
ColoredBall cb = b;

A ColoredBall is a Ball, a Ball is not necessarily a ColoredBall.

Hope you're not too confused.  
This is the hardest part of OOP (objected Oriented Programming) to get.
